File: android/app/src/main/kotlin/dev/agentdeck/terrarium/environment/KelpField.kt
package dev.agentdeck.terrarium.environment

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke
import dev.agentdeck.terrarium.TerrariumColors
import dev.agentdeck.terrarium.TerrariumTiming
import kotlin.math.PI
import kotlin.math.sin

/**
 * Swaying kelp bezier curves growing from the bottom.
 * Each strand has a phase-offset sine for organic movement.
 */
class KelpField {

    private data class KelpStrand(
        val baseX: Float,    // 0..1 fraction of width
        val height: Float,   // 0..1 fraction of canvas height
        val phase: Float,    // sine phase offset
        val segments: Int,   // number of bezier segments (2-4)
    )

    private val strands = listOf(
        KelpStrand(0.08f, 0.25f, 0f, 3),
        KelpStrand(0.12f, 0.30f, 1.2f, 4),
        KelpStrand(0.15f, 0.20f, 2.5f, 2),
        KelpStrand(0.88f, 0.22f, 0.8f, 3),
        KelpStrand(0.92f, 0.28f, 1.8f, 3),
        KelpStrand(0.55f, 0.18f, 3.0f, 2),
    )

    private var time by mutableFloatStateOf(0f)

    fun update(dt: Float) {
        time += dt * TerrariumTiming.KELP_SWAY_SPEED
    }

    fun draw(scope: DrawScope) {
        val w = scope.size.width
        val h = scope.size.height

        for (strand in strands) {
            drawStrand(scope, strand, w, h)
        }
    }

    private fun drawStrand(scope: DrawScope, strand: KelpStrand, w: Float, h: Float) {
        val baseX = strand.baseX * w
        val baseY = h * (1f - 0.02f) // just above bottom
        val topY = baseY - strand.height * h
        val segHeight = (baseY - topY) / strand.segments

        val path = Path().apply {
            moveTo(baseX, baseY)

            for (i in 0 until strand.segments) {
                val sway = sin(time + strand.phase + i * 0.8f) * w * 0.015f * (i + 1)
                val y1 = baseY - (i + 0.5f) * segHeight
                val y2 = baseY - (i + 1f) * segHeight
                val cpX = baseX + sway

                quadraticBezierTo(cpX, y1, baseX + sway * 0.6f, y2)
            }
        }

        // Main stem
        scope.drawPath(
            path = path,
            color = TerrariumColors.KelpDark,
            style = Stroke(width = w * 0.004f, cap = StrokeCap.Round),
        )

        // Lighter inner stroke
        scope.drawPath(
            path = path,
            color = TerrariumColors.KelpGreen.copy(alpha = 0.5f),
            style = Stroke(width = w * 0.002f, cap = StrokeCap.Round),
        )

        // Leaf blobs at segment joints
        for (i in 1..strand.segments) {
            val sway = sin(time + strand.phase + i * 0.8f) * w * 0.015f * i
            val leafY = baseY - i * segHeight
            val leafX = baseX + sway * 0.6f

            scope.drawOval(
                color = TerrariumColors.KelpGreen.copy(alpha = 0.4f),
                topLeft = Offset(leafX - w * 0.006f, leafY - w * 0.003f),
                size = androidx.compose.ui.geometry.Size(w * 0.012f, w * 0.006f),
            )
        }
    }
}

File: apple/AgentDeck/Terrarium/Environment/KelpField.swift
// KelpField.swift — Swaying kelp bezier curves + grass blades
// Ported from android KelpField.kt

import SwiftUI

final class KelpField {
    private struct KelpStrand {
        let baseX: Float
        let height: Float
        let phase: Float
        let segments: Int
    }

    private struct GrassBlade {
        let baseX: Float
        let height: Float
        let phase: Float
        let width: Float
    }

    private let strands: [KelpStrand] = [
        KelpStrand(baseX: 0.08, height: 0.25, phase: 0, segments: 3),
        KelpStrand(baseX: 0.12, height: 0.30, phase: 1.2, segments: 4),
        KelpStrand(baseX: 0.15, height: 0.20, phase: 2.5, segments: 2),
        KelpStrand(baseX: 0.88, height: 0.22, phase: 0.8, segments: 3),
        KelpStrand(baseX: 0.92, height: 0.28, phase: 1.8, segments: 3),
        KelpStrand(baseX: 0.55, height: 0.18, phase: 3.0, segments: 2),
    ]

    private let grassBlades: [GrassBlade] = [
        // Left cluster (near left rocks)
        GrassBlade(baseX: 0.04, height: 0.035, phase: 0.3, width: 0.8),
        GrassBlade(baseX: 0.06, height: 0.050, phase: 1.1, width: 1.0),
        GrassBlade(baseX: 0.08, height: 0.040, phase: 2.0, width: 0.7),
        GrassBlade(baseX: 0.10, height: 0.055, phase: 0.7, width: 0.9),
        GrassBlade(baseX: 0.13, height: 0.030, phase: 1.5, width: 0.8),
        // Center cluster
        GrassBlade(baseX: 0.42, height: 0.045, phase: 2.3, width: 0.9),
        GrassBlade(baseX: 0.44, height: 0.060, phase: 0.5, width: 1.0),
        GrassBlade(baseX: 0.46, height: 0.040, phase: 1.8, width: 0.8),
        GrassBlade(baseX: 0.48, height: 0.050, phase: 3.1, width: 0.7),
        GrassBlade(baseX: 0.43, height: 0.035, phase: 2.8, width: 0.9),
        // Right cluster
        GrassBlade(baseX: 0.83, height: 0.040, phase: 0.9, width: 0.8),
        GrassBlade(baseX: 0.86, height: 0.055, phase: 2.2, width: 1.0),
        GrassBlade(baseX: 0.88, height: 0.035, phase: 1.4, width: 0.7),
        GrassBlade(baseX: 0.90, height: 0.050, phase: 3.5, width: 0.9),
        GrassBlade(baseX: 0.91, height: 0.030, phase: 0.2, width: 0.8),
    ]

    private var time: Float = 0

    func update(dt: Float) {
        time += dt * TerrariumTiming.kelpSwaySpeed
    }

    func draw(context: inout GraphicsContext, size: CGSize) {
        let w = Float(size.width)
        let h = Float(size.height)

        // Draw grass blades first (below kelp)
        for blade in grassBlades {
            drawGrassBlade(context: &context, blade: blade, w: w, h: h)
        }

        for strand in strands {
            drawStrand(context: &context, strand: strand, w: w, h: h)
        }
    }

    private func drawGrassBlade(context: inout GraphicsContext, blade: GrassBlade, w: Float, h: Float) {
        let baseX = blade.baseX * w
        let baseY = h * (1 - TerrariumLayout.sandHeightFraction)
        let tipY = baseY - blade.height * h
        let sway = sin(time * 1.5 + blade.phase) * w * 0.008

        var path = Path()
        path.move(to: CGPoint(x: CGFloat(baseX), y: CGFloat(baseY)))
        path.addQuadCurve(
            to: CGPoint(x: CGFloat(baseX + sway * 0.7), y: CGFloat(tipY)),
            control: CGPoint(x: CGFloat(baseX + sway), y: CGFloat((baseY + tipY) * 0.5))
        )

        context.stroke(path,
                       with: .color(TerrariumColors.kelpDark.opacity(0.6)),
                       lineWidth: CGFloat(w * 0.002 * blade.width))
    }

    private func drawStrand(context: inout GraphicsContext, strand: KelpStrand, w: Float, h: Float) {
        let baseX = strand.baseX * w
        let baseY = h * (1 - TerrariumLayout.sandHeightFraction)
        let topY = baseY - strand.height * h
        let segHeight = (baseY - topY) / Float(strand.segments)

        var path = Path()
        path.move(to: CGPoint(x: CGFloat(baseX), y: CGFloat(baseY)))
        for i in 0..<strand.segments {
            let sway = sin(time + strand.phase + Float(i) * 0.8) * w * 0.015 * Float(i + 1)
            let y1 = baseY - (Float(i) + 0.5) * segHeight
            let y2 = baseY - (Float(i) + 1) * segHeight
            let cpX = baseX + sway
            path.addQuadCurve(
                to: CGPoint(x: CGFloat(baseX + sway * 0.6), y: CGFloat(y2)),
                control: CGPoint(x: CGFloat(cpX), y: CGFloat(y1))
            )
        }

        // Main stem (dark)
        context.stroke(path,
                       with: .color(TerrariumColors.kelpDark),
                       lineWidth: CGFloat(w * 0.004))

        // Lighter inner stroke
        context.stroke(path,
                       with: .color(TerrariumColors.kelpGreen.opacity(0.5)),
                       lineWidth: CGFloat(w * 0.002))

        // Leaf blobs at segment joints
        for i in 1...strand.segments {
            let sway = sin(time + strand.phase + Float(i) * 0.8) * w * 0.015 * Float(i)
            let leafY = baseY - Float(i) * segHeight
            let leafX = baseX + sway * 0.6

            let rect = CGRect(x: CGFloat(leafX - w * 0.006), y: CGFloat(leafY - w * 0.003),
                              width: CGFloat(w * 0.012), height: CGFloat(w * 0.006))
            context.fill(Path(ellipseIn: rect),
                         with: .color(TerrariumColors.kelpGreen.opacity(0.4)))
        }
    }
}
